sdk/eventhub/event-hubs/src/eventHubConsumerClient.ts

```diff
@@ -320,9 +379,96 @@ export class EventHubConsumerClient {
       get isRunning() {
         return eventProcessor.isRunning();
       },
-      close: () => eventProcessor.stop()
+      close: () => {
+        this._partitionGate.remove(targetedPartitionId);
+        return eventProcessor.stop();
+      }
     };
   }
+
+  private createEventProcessorForAllPartitions(
+    subscriptionEventHandlers: SubscriptionEventHandlers,
+    options?: SubscribeOptions
+  ) {
+    this._partitionGate.add("all");
+
+    if (this._userChoseCheckpointStore) {
+      log.consumerClient("Subscribing to all partitions, using a checkpoint store.");
+    } else {
+      log.consumerClient("Subscribing to all partitions, no checkpoint store.");
+    }
+
+    const subscribeOptions = options as SubscribeOptions | undefined;
+
+    const eventProcessor = this._createEventProcessor(
+      this._consumerGroup,
+      this._eventHubClient,
+      subscriptionEventHandlers,
+      this._checkpointStore,
+      {
+        ...defaultConsumerClientOptions,
+        ...(options as SubscribeOptions),
+        ownerLevel: getOwnerLevel(subscribeOptions, this._userChoseCheckpointStore),
+        partitionLoadBalancer: this._userChoseCheckpointStore
+          ? undefined
+          : new GreedyPartitionLoadBalancer()
+      }
+    );
+
+    return { targetedPartitionId: "all", eventProcessor };
+  }
+
+  private createEventProcessorForSinglePartition(
+    partitionId: string,
+    eventHandlers: SubscriptionEventHandlers,
+    options?: SubscribeOptions
+  ) {
+    this._partitionGate.add(partitionId);
+
+    const subscribeOptions = options as SubscribeOptions | undefined;
+    
+    if (this._userChoseCheckpointStore) {
+      log.consumerClient(
+        `Subscribing to specific partition (${partitionId}), using a checkpoint store`
+      );      
+    } else {
+      log.consumerClient(
+        `Subscribing to specific partition (${partitionId}), no checkpoint store.`
+      );
+    }
+
+    const eventProcessor = this._createEventProcessor(
+      this._consumerGroup,
+      this._eventHubClient,
+      eventHandlers,
+      this._checkpointStore,
+      {
+        ...defaultConsumerClientOptions,
+        ...options,
+        // this load balancer will just grab _all_ the partitions, not looking at ownership
+        partitionLoadBalancer: new GreedyPartitionLoadBalancer([partitionId]),
```

Comment:
Remind me...did we decide to forego claiming partitions if we call subscribe on a single partition with a checkpointstore? What happens if I am running my app on 2 different processes and so there are 2 EventProcessors trying to read from the same partition with an ownerLevel? Will the GreedyPartitionLoadBalancer cause them to continuously steal from each other?

I wouldn't expect the above to actually happen outside of some niche use-cases.

Reply:
I don't think they can actually get into that situation. 

The user can't choose a load balancer so they're going to go into one of two buckets - either they specify _nothing_ and they get the greedy load balancer (and no durable checkpointing) or they specified a checkpointstore (in which we case give them the FairPartitionLoadBalancer).

Reply:
If the user has passed a checkpointstore into the EventHubConsumerClient constructor, then calls subscribe with a partitionId, `createEventProcessorForSinglePartition` gets invoked:
https://github.com/Azure/azure-sdk-for-js/pull/6304/files/cf6f1287c084500f4fcb610228b199dc4c389942#diff-b6b3d7605c67e8cc8286917a7e2f030dR367

In this line, you're hardcoding the GreedyPartitionLoadBalancer, so they'll never get a FairLoadBalancer.

When `loadBalance` gets called, you'll first get the list of potential partition ids to claim:

https://github.com/Azure/azure-sdk-for-js/blob/cf6f1287c084500f4fcb610228b199dc4c389942/sdk/eventhub/event-hubs/src/partitionLoadBalancer.ts#L51-L56
Since `partitionsToClaim` contains the partitionId you passed in the constructor, you'll get back that same array.

Then further down in `loadBalance`, you filter out partitionIds that are in the `partitionOwnershipMap`.
https://github.com/Azure/azure-sdk-for-js/blob/cf6f1287c084500f4fcb610228b199dc4c389942/sdk/eventhub/event-hubs/src/partitionLoadBalancer.ts#L68

Now consider the case now where you've called subscribe on a single partition and have been checkpointing. Eventually your subscribe call ends (either the process crashed, or maybe you called stop yourself). Now later on you call subscribe again with the same partition id. You'll still have an entry for the partition in the partition ownership map (since it's persistent... I passed in a checkpointstore), and loadBalance will return an empty list. At this point you can no longer call subscribe on a partition id you've already subscribed to in the past. (Well, you can, you just won't receive anything)

Reply:
Ah, okay. Let's split this into two smaller cases:

Case #1 - user stops the subscriber - this is handled by #6333 which will take care of "abandoning" ownership when we gracefully drop off.

Case #2 - things don't exit cleanly. This one we should fix. The simplest thing I can think of is to just keep track of our event processor ID (much like the fair partition load balancer does) and just obliterate any ownership for partitions that we're asked to own in the greedy load balancer (which is basically it's whole reason for existing). I'll take a look at this tomorrow - I don't think it'll be too difficult.

Reply:
I added in some code for case #2:
* GreedyPartitionLoadBalancer will now steal any partitions that it doesn't own (where own is defined as "has a different ownerID than mine"
* EventHubConsumerClient now chooses the ID for the processors rather than letting EventHubProcessor do it. This lets us maintain a consistent ID for _all_ subscriptions tha are happening from the same EventHubConsumerClient.

I'm testing out this approach now - I've pushed the code if you want to take a look at it.